### PR TITLE
Implementation of Gaussian, GaussianMixture, ParticleSet and related

### PR DESCRIPTION
**Pull request by xEnVrE**
Within this PR:
- Class `GaussianMixture` has now methods to access to mean, covariance and weight of the i-th component
- Class `Gaussian` re-implemented as inheriting class from class `GaussianMixture`
- Class `GaussianRef` removed
- Implemented class `ParticleSet` as inheriting class from class `GaussianMixture`
- Use class `ParticleSet` within classes `PFPrediction`, `PFPredictionDecorator`, `PFCorrection`, `PFCorrectionDecorator`, `DrawParticles`, `UpdateParticles`, `Resampling`, `ResamplingWithPrior`, `ParticleSetInitialization`, `InitSurveillanceAreaGrid` and `SIS`
- Method `ResamplingWithPrior::resample` heavily changed (due to use of ParticleSet)
- Method `SIS::SIS` take the state size as argument (required to initialize ParticleSet)
- Update `test_Gaussian`, `test_SigmaPointUtils `(since use class Gaussian)
- Update `test_SIS` (since use class `SIS`)
- Update `test_SIS_Decorators` (since use classes `PFPredictionDecorator`, `PFCorrectionDecorator` and `SIS`)
- Increase version to `0.8.101` since breaking API
- Update `CHANGELOG.md`

### CODE REVIEW

**Comment by claudiofantacci** [comment]
@xEnVrE meanwhile I'm reviewing this PR, can you please rebase onto devel to update Travis script?

**Comment by xEnVrE** [comment]
@claudiofantacci rebased on last devel

File: src/BayesFilters/include/BayesFilters/Gaussian.h
Comment (claudiofantacci) [inline src/BayesFilters/include/BayesFilters/Gaussian.h]: For comments, please use the following format:
```cpp
/**
 * First line here
 * Next line here
 * Final line here 
 */
```

File: test/test_Gaussian/main.cpp
Comment (claudiofantacci) [inline test/test_Gaussian/main.cpp]: ~~I don't really like this syntax, but let's keep it like this for now.~~
~~I suggest to add element-wise accessor to `GaussianMixture` and `Gaussian`, for example.~~
```cpp
gaussian_3d(0) = 2;
```
~~implies the first element of the mean vector;~~
```cpp
gaussian_mixture_3d(3, 0) = 2;
```
~~implies the first element of the mean vector of the 3rd gaussian component of the mixture.~~
~~Same reasoning with `ParticleSet`/`Particle`.~~

~~I would also do the same for the covariances.~~

File: test/test_Gaussian/main.cpp
Comment (claudiofantacci) [inline test/test_Gaussian/main.cpp]: Please remove the `catch_expetion` variable and move the `return EXIT_FAILURE` inside the `catch` block.

File: src/BayesFilters/include/BayesFilters/ParticleSet.h
Comment (claudiofantacci) [inline src/BayesFilters/include/BayesFilters/ParticleSet.h]: I would suggest to also implement `operator+=` and to follow guidelines reported [here](https://en.cppreference.com/w/cpp/language/operators).

File: src/BayesFilters/src/ResamplingWithPrior.cpp
Comment (claudiofantacci) [inline src/BayesFilters/src/ResamplingWithPrior.cpp]: Missing space after `:`.

**Review by claudiofantacci — CHANGES_REQUESTED** [review]
LGTM!

Thanks for the very good PR.
I have only minor comments and suggestions.
After addressing them, I'll merge the PR.

Should you need to add/remove/modify anything within this PR, remember to update the `CHANGELOG`.

File: src/BayesFilters/src/ResamplingWithPrior.cpp
Comment (claudiofantacci) [inline src/BayesFilters/src/ResamplingWithPrior.cpp]: Change `res_parents` type to `VectorXi`.

File: src/BayesFilters/include/BayesFilters/ParticleSet.h
Comment (xEnVrE) [reply src/BayesFilters/include/BayesFilters/ParticleSet.h]: Done, however I decided not to put the symmetric operator

`ParticleSet operator+(ParticleSet lhs, const ParticleSet& rhs)`

as a member function because in that case it is required
- to put the definition of the function within the class body, i.e., within the header file
- to declare it as friend (but in this case friendship is not required)

File: test/test_Gaussian/main.cpp
Comment (xEnVrE) [reply test/test_Gaussian/main.cpp]: Implemented accessors for `Gaussian` and `GaussianMixture` (both mean and covariance).

As regards `ParticleSet` we should decide how to disambiguate between element-wise access to state and mean.

**Comment by xEnVrE** [comment]
Class `Particle` is not used at all and I left the original implementation as in the latest `devel`. Should we remove the files associated to it?

File: test/test_Gaussian/main.cpp
Comment (claudiofantacci) [reply test/test_Gaussian/main.cpp]: I correct my previous examples that are wrong! I apologize.

I don't really like this syntax, but let's keep it like this for now.
I suggest to add element-wise accessor to `GaussianMixture` and `Gaussian`, for example.
```cpp
gaussian_3d.mean(0) = 2;
gaussian_3d.covariance(0, 0) = 2;
```
implies the first element of the mean vector and the `(0, 0)` element of the covariance;
```cpp
gaussian_mixture_3d.mean(3, 0) = 2;
gaussian_mixture_3d.covariance(3, 0, 0) = 2;
```
implies the first element of the mean vector and the `(0, 0)` element of the covariance of the 3rd gaussian component of the mixture.
Same reasoning with `ParticleSet`.

I would also do the same for the covariances.

**Comment by claudiofantacci** [comment]
> Class `Particle` is not used at all and I left the original implementation as in the latest `devel`. Should we remove the files associated to it?

Let's delete it!

File: test/test_Gaussian/main.cpp
Comment (xEnVrE) [reply test/test_Gaussian/main.cpp]: Now it should be ok.

I also had to re-implement `Gaussian::covariance()` since the new method `Gaussian::covariance(const std:size_t i, const std::size_t j)` hides `GaussianMixture::covariance()` which was ok also for a Gaussian since the matrix of covariances shrinks to a single covariance matrix.

As regards `ParticleSet`, the methods for mean/covariance are inherited from `GaussianMixture` while I implemented element-wise access to the i-th state.

File: test/test_Gaussian/main.cpp
Comment (claudiofantacci) [inline test/test_Gaussian/main.cpp]: Almost there!
Please delete this line and `catch_exception` declaration some lines above as we are no longer using it.

After this, I merge the PR 🎉

**Review by claudiofantacci — CHANGES_REQUESTED** [review]

File: test/test_Gaussian/main.cpp
Comment (xEnVrE) [reply test/test_Gaussian/main.cpp]: Sorry, forgot to remove them!

**Review by claudiofantacci — APPROVED** [review]

**Comment by claudiofantacci** [comment]
LGTM!

Thanks for the big effort and contribution @xEnVrE!

**Pull request merged**
